### PR TITLE
Add txpower and frame port support to TinyLoRa. Modify only one example.

### PR DESCRIPTION
**Pull request by clavisound**
Hello there,

Added support for frame port and txpower. I tested only frame port and it works.

I modified **only one** arduino (.ino) example accordingly.

If you like my changes, I will try to correct also the other examples. It's my first time in C and OO, so please review my changes if you agree to include them.

Binary size is exactly the same for 32u4.

(I also added a missing spacebar character)

### CODE REVIEW

**Comment by clavisound** [comment]
Hmmm, pass failed because of dht example? I paste the modified version.

Here's the diff
```
diff -u ../libraries/TinyLoRa/examples/tinylora_dht22/tinylora_dht22.ino tinylora_dht22-clv.ino 
--- ../libraries/TinyLoRa/examples/tinylora_dht22/tinylora_dht22.ino	2019-08-07 16:29:45.000000000 +0300
+++ tinylora_dht22-clv.ino	2019-08-29 15:53:07.919917934 +0300
@@ -26,6 +26,10 @@
 uint8_t DevAddr[4] = { 0x00, 0x00, 0x00, 0x00 };
 
 /************************** Example Begins Here ***********************************/
+// setup TX settings
+uint8_t const TxPower = 0xff; // Default 0xff.
+uint8_t FramePort = 1;
+
 // Data Packet to Send to TTN
 unsigned char loraData[4];
 
@@ -57,7 +61,7 @@
   lora.setChannel(MULTI);
   // set datarate
   lora.setDatarate(SF7BW125);
-  if(!lora.begin())
+  if(!lora.begin(TxPower))
   {
     Serial.println("Failed");
     Serial.println("Check your radio");
@@ -103,7 +107,7 @@
   loraData[3] = lowByte(humidInt);
   
   Serial.println("Sending LoRa Data...");
-  lora.sendData(loraData, sizeof(loraData), lora.frameCounter);
+  lora.sendData(loraData, sizeof(loraData), lora.frameCounter, FramePort);
   Serial.print("Frame Counter: ");Serial.println(lora.frameCounter);
   lora.frameCounter++;
```

Full score :-)
```
// TinyLoRa DHT22 - ABP TTN Packet Sender (Multi-Channel)
// Tutorial Link: https://learn.adafruit.com/the-things-network-for-feather/using-a-feather-32u4
//
// Adafruit invests time and resources providing this open source code.
// Please support Adafruit and open source hardware by purchasing
// products from Adafruit!
//
// Copyright 2015, 2016 Ideetron B.V.
//
// Modified by Brent Rubell for Adafruit Industries, 2018
/************************** Configuration ***********************************/
#include <TinyLoRa.h>
#include <SPI.h>
#include "DHT.h"

// Visit your thethingsnetwork.org device console
// to create an account, or if you need your session keys.

// Network Session Key (MSB)
uint8_t NwkSkey[16] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };

// Application Session Key (MSB)
uint8_t AppSkey[16] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };

// Device Address (MSB)
uint8_t DevAddr[4] = { 0x00, 0x00, 0x00, 0x00 };

/************************** Example Begins Here ***********************************/
// setup TX settings
uint8_t const TxPower = 0xff; // Default 0xff.
uint8_t FramePort = 1;

// Data Packet to Send to TTN
unsigned char loraData[4];

// How many times data transfer should occur, in seconds
const unsigned int sendInterval = 30;

// Pinout for Adafruit Feather 32u4 LoRa
TinyLoRa lora = TinyLoRa(7, 8, 4);

// Pinout for Adafruit Feather M0 LoRa
//TinyLoRa lora = TinyLoRa(3, 8, 4);

// pin the DHT22 is connected to
#define DHTPIN 10
DHT dht(DHTPIN, DHT22);

void setup()
{
  delay(2000);
  Serial.begin(9600);
  while (! Serial);
 
  // Initialize pin LED_BUILTIN as an output
  pinMode(LED_BUILTIN, OUTPUT);
  
  // Initialize LoRa
  Serial.print("Starting LoRa...");
  // define multi-channel sending
  lora.setChannel(MULTI);
  // set datarate
  lora.setDatarate(SF7BW125);
  if(!lora.begin(TxPower))
  {
    Serial.println("Failed");
    Serial.println("Check your radio");
    while(true);
  }
  Serial.println("OK");

  // Initialize DHT
  dht.begin();
}

void loop()
{
   // Reading temperature or humidity takes about 250 milliseconds!
  // Sensor readings may also be up to 2 seconds 'old' (its a very slow sensor)
  float h = dht.readHumidity();
  // Read temperature as Celsius (the default)
  float t = dht.readTemperature();

  Serial.print("Humidity: ");
  Serial.print(h);
  Serial.print(" %\t");
  Serial.print("Temperature: ");
  Serial.print(t);
  Serial.println("");
  
  // Check if any reads failed and exit early (to try again).
  if (isnan(h) || isnan(t)) {
    Serial.println("Failed to read from DHT sensor!");
    return;
  }

  // encode float as int
  int16_t tempInt = round(t * 100);
  int16_t humidInt = round(h * 100);

  // encode int as bytes
  //byte payload[2];
  loraData[0] = highByte(tempInt);
  loraData[1] = lowByte(tempInt);
  
  loraData[2] = highByte(humidInt);
  loraData[3] = lowByte(humidInt);
  
  Serial.println("Sending LoRa Data...");
  lora.sendData(loraData, sizeof(loraData), lora.frameCounter, FramePort);
  Serial.print("Frame Counter: ");Serial.println(lora.frameCounter);
  lora.frameCounter++;

  // blink LED to indicate packet sent
  digitalWrite(LED_BUILTIN, HIGH);
  delay(1000);
  digitalWrite(LED_BUILTIN, LOW);
  
  Serial.println("delaying...");
  delay(sendInterval * 1000);
}
```

File: examples/hello_LoRa/hello_LoRa.ino [hunk 1]
```diff
@@ -25,6 +25,10 @@ uint8_t AppSkey[16] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x
 uint8_t DevAddr[4] = { 0x00, 0x00, 0x00, 0x00 };
 
 /************************** Example Begins Here ***********************************/
+// setup TX settings
+uint8_t const TxPower = 0xff; // Default 0xff.
+uint8_t FramePort = 2;
```
Comment (clavisound) [inline examples/hello_LoRa/hello_LoRa.ino]: please change `2` to `1`. I was testing and I forgot it to `2`

File: examples/hello_LoRa/hello_LoRa.ino [hunk 2]
```diff
@@ -52,7 +56,7 @@ void setup()
   lora.setChannel(MULTI);
   // set datarate
   lora.setDatarate(SF7BW125);
-  if(!lora.begin())
+  if(!lora.begin(TxPower))
```
Comment (brentru) [inline examples/hello_LoRa/hello_LoRa.ino]: I don't think we should modify `begin()`. Instead, could you add a `lora.setPower()` function which performs `RFM_Write(0x09, TX_POWER);`, like how we have `setChannel` and `setDatarate`?

File: examples/hello_LoRa/hello_LoRa.ino (hunk 1 above)
Comment (brentru) [reply examples/hello_LoRa/hello_LoRa.ino]: Please make changes to this pull request to set this example to `1`?

File: TinyLoRa.cpp [hunk 3]
```diff
@@ -520,7 +522,7 @@ void TinyLoRa::sendData(unsigned char *Data, unsigned char Data_Length, unsigned
   unsigned char Mac_Header = 0x40;
 
   unsigned char Frame_Control = 0x00;
-  unsigned char Frame_Port = 0x01;
+//  unsigned char Frame_Port = 0x01;
```
Comment (brentru) [inline TinyLoRa.cpp]: Please do not comment this out. Instead, the code should default `Frame_Port` to `0x01` unless `Frame_Port` is passed into `sendData`.

File: TinyLoRa.cpp [hunk 4]
```diff
@@ -51,6 +51,8 @@
 extern uint8_t NwkSkey[16]; ///< Network Session Key
 extern uint8_t AppSkey[16]; ///< Application Session Key
 extern uint8_t DevAddr[4]; ///< Device Address
+extern uint8_t Tx_Power; ///< Tx power
```
Comment (brentru) [inline TinyLoRa.cpp]: Why are you setting `Tx_Power` and `Frame_Port` here?

**Review by brentru — CHANGES_REQUESTED** [review]
Thanks for the pull request! I'd like to suggest a few changes to the code before we'd merge this in. Let's make these changes first, and then I'll take another look at the build script.

File: examples/hello_LoRa/hello_LoRa.ino (hunk 2 above)
Comment (clavisound) [reply examples/hello_LoRa/hello_LoRa.ino]: I agree, I will try to do it. I can't promise result.

File: TinyLoRa.cpp (hunk 4 above)
Comment (clavisound) [reply TinyLoRa.cpp]: I don't know a better place! It's useless? I have to moved / removed?

File: examples/hello_LoRa/hello_LoRa.ino (hunk 1 above)
Comment (clavisound) [reply examples/hello_LoRa/hello_LoRa.ino]: I didn't knew this was possible. Thanks.

File: examples/hello_LoRa/hello_LoRa.ino (hunk 2 above)
Comment (clavisound) [reply examples/hello_LoRa/hello_LoRa.ino]: If removed from begin(). Begin should call the lora.setPower function to have a proper init of device?

File: examples/hello_LoRa/hello_LoRa.ino (hunk 1 above)
Comment (brentru) [reply examples/hello_LoRa/hello_LoRa.ino]: Yup - check out https://learn.adafruit.com/contribute-to-circuitpython-with-git-and-github/receiving-a-review

File: examples/hello_LoRa/hello_LoRa.ino (hunk 2 above)
Comment (brentru) [reply examples/hello_LoRa/hello_LoRa.ino]: I'm not sure, but you _might_ be able to set the PA independently of the begin function.

File: TinyLoRa.cpp (hunk 4 above)
Comment (brentru) [reply TinyLoRa.cpp]: You might want to place it here: https://github.com/adafruit/TinyLoRa/blob/master/TinyLoRa.h#L105

File: examples/hello_LoRa/hello_LoRa.ino (hunk 1 above)
Comment (brentru) [reply examples/hello_LoRa/hello_LoRa.ino]: Let's remove these from the example, making a separate example showing increasing the TX power before sending:
```
void loop()
{
  Serial.println("Sending LoRa Data...");
  lora.setTxPower(0xff);
  lora.sendData(loraData, sizeof(loraData), lora.frameCounter);
...
```

Similar to this method in our RFM9x CircuitPython library: https://github.com/adafruit/Adafruit_CircuitPython_RFM9x/blob/master/adafruit_rfm9x.py#L528

File: TinyLoRa.cpp (hunk 4 above)
Comment (clavisound) [reply TinyLoRa.cpp]: Removed Frame_Port, moved Tx_Power, I hope it's ok now.

File: examples/hello_LoRa/hello_LoRa.ino (hunk 2 above)
Comment (clavisound) [reply examples/hello_LoRa/hello_LoRa.ino]: If made some tests: if `lora.setPower` before `begin()` feather 32u4 hangs. But seems safe after begin() (init).

But now I have another issue. It seems - **if** I understood correctly - that according to [HOPE RFM9x documentation, p.80, section 5.4.3](https://www.hoperf.com/data/upload/portal/20190801/RFM95W-V2.0.pdf), high power setting must erased if PA_BOOST is NOT selected but PA_LF / PA_HF choosed.

I handle this in .ino. Better to move it to library the checking of values? (if needed)

File: examples/hello_LoRa/hello_LoRa.ino [hunk 5]
```diff
@@ -25,6 +25,33 @@ uint8_t AppSkey[16] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x
...
+uint8_t const TxPower = 0xFF; // default value. Don't modify this value [default: 0xFF]
+
+// uncomment this section if you want to control the TxPower, and comment the above line.
+/*
+bool PaBoost = 1; // 8th bit [default: 1]
+uint8_t OutputPower = 1; // 1-4 bit (DEC values: 0-15) [default: 15] +2 to +17dBm.
+uint8_t MaxPower = 4; // 5-7 bit, (DEC values: 0-7) [default: 4] BUG? Range -4 to 0 but DOC says -4 to +15dBm
+
+// function to pack the data for TXpower.
+uint8_t packDataPower(){
```
Comment (blackrosezy) [inline examples/hello_LoRa/hello_LoRa.ino:38]: I think all calculation should be inside the lora.setPower(TxPower) function.

Comment (clavisound) [reply examples/hello_LoRa/hello_LoRa.ino:38]: It seems the setPower (aka the function in .ino) is not ready yet, although I tested and it works. I managed to get -100 to -9 RSSI from 30 cm distance. I think some details are missing. I am getting some ideas from lmic. https://github.com/mcci-catena/arduino-lmic/commit/49b495def976212b954b0af3904ae492d8f468da

Is preferable to commit in this request, or start new?

File: examples/hello_LoRa/hello_LoRa.ino (hunk 1 above)
Comment (clavisound) [reply examples/hello_LoRa/hello_LoRa.ino]: I wrote also a dBm function, I can put it to the new example.

File: examples/hello_LoRa/hello_LoRa.ino (hunk 5 above)
Comment (brentru) [reply examples/hello_LoRa/hello_LoRa.ino:38]: @clavisound Could you make a new PR specifically for `setPower` when you're ready?

File: examples/hello_LoRa/hello_LoRa.ino (hunk 2 above)
Comment (brentru) [reply examples/hello_LoRa/hello_LoRa.ino]: @clavisound TinyLoRa doesn't support configuring high power (+20dBm) mode, but it'd be great to add it!

The power mode (high power, `RegPaDac`) would be a boolean passed into the `begin()` call: `lora.begin(powerMode)`

Seeing as the Feather 32u4 hangs if power is set before initialization, place the `setPower` call after the `begin()`.

The `setPower` function checks if high power was selected during initialization.

The .ino should only have the function calls `begin`/`setpower` and not checks, those'd be handled by the library

File: examples/hello_LoRa/hello_LoRa.ino (hunk 5 above)
Comment (clavisound) [reply examples/hello_LoRa/hello_LoRa.ino:38]: Yes, I had the same idea. I came to write you.

I will submit for frame port only. I will configure the default port to 1, so the examples will not brake and I will try to write separate example for frame port.

You can deny this PR.

Thank you both for the time.

Comment (brentru) [reply examples/hello_LoRa/hello_LoRa.ino:38]: Sounds good, closing this PR and I'm excited to test your next PR.

**Pull request closed without merging**